### PR TITLE
[Regular Expressions] simplify duplicate group start logic

### PR DESCRIPTION
**Pull request by keith-hall**
builds on the excellent work done in https://github.com/sublimehq/Packages/pull/3054 (so that one can be merged first), and removes the need for duplicating all the match patterns inside `group-start` into `group-start-extended`. Thus it should be easier to create syntaxes extending the basic one.

### CODE REVIEW

**Comment by deathaxe** [comment]
Merging the other PR first is likely to cause conflicts due to squashing.

I already had a feeling about deduplication becomming possible due to those changes. Thanks for proofing!

File: Regular Expressions/RegExp (Basic).sublime-syntax [hunk 1]
```diff
@@ -90,84 +86,123 @@ contexts:
...
-      set: [group-body, maybe-unexpected-quantifiers]
     - include: group-start-common
-    - match: ''
-      set: [group-body, maybe-unexpected-quantifiers]
 
   group-start-extended:
     - meta_scope: meta.group.extended.regexp
+    - include: group-start-common
+
+  group-start-common:
```
Comment (deathaxe) [inline Regular Expressions/RegExp (Basic).sublime-syntax:114]: Can't make a github suggestion for such a large block, but how about...

```
  groups:
    - match: \(
      scope: meta.group.regexp keyword.control.group.regexp
      push: [group-body, maybe-unexpected-quantifiers, group-start]

  groups-extended:
    - match: \(
      scope: meta.group.extended.regexp keyword.control.group.regexp
      push: [group-body-extended, maybe-unexpected-quantifiers, group-start]

  group-start:
     - match: \?(<[=!]|[>=:!])
```

By adding the meta scope to the opening `\(` match it appears to be possible to remove some of those intermediate contexts. Note `group-start-common` becomes `group-start`.

File: Regular Expressions/RegExp (Basic).sublime-syntax [hunk 2]
```diff
@@ -90,84 +86,123 @@ contexts:
...
-    - match: '\?(?:[ixms]*-)?[ims]+:'
+    - match: '\?(?:[ixms]*-)?[ixms]+:'
       scope: storage.modifier.mode.regexp
-      set: [group-body-extended, maybe-unexpected-quantifiers]
-    - include: group-start-common
-    - match: ''
-      set: [group-body-extended, maybe-unexpected-quantifiers]
-
-  group-start-common:
+      pop: 2
```
Comment (deathaxe) [inline Regular Expressions/RegExp (Basic).sublime-syntax:131]: Wouldn't we just want to continue with `maybe-unexpected-quantifiers` if this pattern matches?

```suggestion
      pop: 1
```

File: Regular Expressions/RegExp (Basic).sublime-syntax [hunk 3]
```diff
@@ -90,84 +86,123 @@ contexts:
...
     # Other modifiers (must come after other modifier matches)
     - match: (\?(?:[ixms]*-)?[ixms]+)(\))
       captures:
         1: storage.modifier.mode.regexp
         2: keyword.control.group.regexp
+      pop: 2
```
Comment (deathaxe) [inline Regular Expressions/RegExp (Basic).sublime-syntax:137]: We propably don't want to end up in `group-body` if a closing paren is already matched?

```suggestion
      pop: 3
```

**Review by deathaxe — APPROVED** [review]

File: Regular Expressions/RegExp (Basic).sublime-syntax [hunk 4]
```diff
@@ -90,84 +87,113 @@ contexts:
...
+        2: storage.modifier.mode.regexp
+        3: keyword.control.group.regexp
+      set: [base-literal-extended, maybe-unexpected-quantifiers]
+    # Deactivates 'x' mode, but already not active
+    - match: (\()({{deactivate_x_mode}})(\))
+      scope: meta.group.regexp
+      captures:
+        1: keyword.control.group.regexp
+        2: storage.modifier.mode.regexp
+        3: keyword.control.group.regexp
```
Comment (deathaxe) [inline Regular Expressions/RegExp (Basic).sublime-syntax:179]: Is that one still needed? If yes, would it need to be copied to `ingroup-xmode-on`?

It seems to be handled by `other_modifiers` in `group-start`.

File: Regular Expressions/RegExp (Basic).sublime-syntax [hunk 5]
```diff
@@ -90,84 +87,113 @@ contexts:
...
+      set: [base-literal, maybe-unexpected-quantifiers]
+    # Activates 'x' mode, but already active
+    - match: (\()({{activate_x_mode}})(\))
+      scope: meta.group.extended.regexp
+      captures:
+        1: keyword.control.group.regexp
+        2: storage.modifier.mode.regexp
+        3: keyword.control.group.regexp
```
Comment (deathaxe) [inline Regular Expressions/RegExp (Basic).sublime-syntax:196]: Same here.

File: Regular Expressions/RegExp (Basic).sublime-syntax (hunk 4 above)
Comment (keith-hall) [reply Regular Expressions/RegExp (Basic).sublime-syntax]: you're quite right, it's not needed here any more :)

File: Regular Expressions/RegExp (Basic).sublime-syntax (hunk 5 above)
Comment (keith-hall) [reply Regular Expressions/RegExp (Basic).sublime-syntax]: removed :)

**Comment by deathaxe** [comment]
Just noticed the current implementation has some issues with correctly stacking `meta.group` scopes if global modifiers exist

Example:

```
(
# <- meta.group.regexp keyword.control.group.regexp
#^ meta.group.regexp - keyword
    (?x)
#  ^ meta.group.regexp
#   ^^^^ meta.group.regexp meta.group.extended.regexp
#   ^ keyword.control.group.regexp
#    ^^ storage.modifier.mode.regexp
#      ^ keyword.control.group.regexp
    # comment
#   ^^^^^^^^^ comment
    (?x)
#   ^^^^ meta.group.extended.regexp meta.group.extended.regexp
#   ^ keyword.control.group.regexp
#    ^^ storage.modifier.mode.regexp
#      ^ keyword.control.group.regexp
    # comment
#   ^^^^^^^^^ comment
    (?-x)
#   ^^^^^ meta.group.extended.regexp meta.group.regexp
    # comment
#   ^^^^^^^^^ - comment
    (?-x)
#   ^^^^^ meta.group.regexp meta.group.regexp
    # comment
#   ^^^^^^^^^ - comment
) # no comment
# <- keyword.control.group
# ^ - comment
```

**Comment by keith-hall** [comment]
What would be the preferred outcome/correct stacking? :)

**Comment by deathaxe** [comment]
The tests in the comment already fail because e.g. first `(?x)` contains only one `meta.group`.

**Comment by keith-hall** [comment]
Ah, understood - sorry, I somehow didn't pick up on that. I have pushed a candidate fix to address this - those tests now pass.

**Review by deathaxe — APPROVED** [review]

**Comment by deathaxe** [comment]
Here's what I have with regards to `meta.mode` so far.

https://github.com/deathaxe/sublime-packages/tree/pr/regexp/meta-xmode

It changes some scopes, so I am not sure whether it should go in here.

**Comment by keith-hall** [comment]
I like your changes, but I personally think it would be a good idea to merge this PR and make new PRs to master afterwards. That way, it will be easier to review and discuss.

**Review by FichteFoll — APPROVED** [review]

**Pull request merged**